src/xwayland/xserver.rs

```diff
@@ -1,509 +1,378 @@
- * STDOUT.
- */
 use std::{
     env,
     ffi::OsStr,
-    io::{self, Read},
+    os::fd::{BorrowedFd, OwnedFd},
     os::unix::{
-        io::{AsRawFd, BorrowedFd, RawFd},
+        io::{AsRawFd, RawFd},
         net::UnixStream,
         process::CommandExt,
     },
-    process::{ChildStdout, Command, Stdio},
+    process::Command,
     sync::{Arc, Mutex},
 };
 
-use calloop::{
-    channel::{self, sync_channel, Channel, SyncSender},
-    generic::Generic,
-    Interest, LoopHandle, Mode,
-};
-use wayland_server::{
-    backend::{ClientData, ClientId, DisconnectReason},
-    Client, DisplayHandle,
-};
+use tracing::{info, trace};
+use wayland_server::backend::ClientData;
+use wayland_server::backend::DisconnectReason;
+use wayland_server::{Client, DisplayHandle};
 
-use tracing::{error, info, instrument};
+use crate::{utils::user_data::UserDataMap, wayland::compositor::CompositorClientState};
 
 use super::x11_sockets::{prepare_x11_sockets, X11Lock};
-use crate::utils::user_data::UserDataMap;
-#[cfg(feature = "wayland_frontend")]
-use crate::wayland::compositor::CompositorClientState;
 
-/// The XWayland handle
+/// A handle to a running XWayland process. Using XWayland as an xserver for
+/// X11-based clients requires two connections: one wayland socket, where
+/// XWayland creates surfaces for its clients, and one X11 socket, where a
+/// compositor's implementation of an X11 window manager connects and handles
+/// X11 events.
+///            ┌───────────────────┐
+///            │                   │
+///            │                   │ X11 Clients            
+///            │     Xwayland      │ ◄─────
+///            │                   │
+///            │                   │
+///            └──┬──────────────▲┘
+///               │              │
+/// Wayland Socket│              │ X11 Socket   
+///               │              │
+///               │              │
+///  ┌────────────▼─────────────┼────────────┐
+///  │  Compositor               │            │
+///  │                         ┌─┴───────┐    │
+///  │                         │ XWM     │    │
+///  │                         │         │    │
+///  │                         └─────────┘    │
+///  │                                        │
+///  └────────────────────────────────────────┘
+///
+/// This struct handles integrating the XWayland process itself into the event
+/// loop, but a [separate X11 window manager implementation](crate::xwayland::xwm::X11Wm)
+/// is needed as well to support X11 clients.
+///
+/// To shut down the instance, dropping this handle is generally sufficient,
+/// along with the X11 connection that was passed to the window manager
+/// implementation, if any. The process will die once the connections to it
+/// are closed.
 #[derive(Debug)]
 pub struct XWayland {
-    inner: Arc<Mutex<Inner>>,
+    inner: Arc<Mutex<Instance>>,
+    source: calloop::generic::Generic<calloop::generic::FdWrapper<RawFd>>,
```

Comment:
Is there a reason OwnedFd can't be used here?

Reply:
Both the handle and the source need the fd, and it made more sense to me for the handle to own it.